hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java

```diff
@@ -0,0 +1,179 @@
+      }
+
+      // Perform the actual upgrade/downgrade; this has to be idempotent, for now.
+      LOG.info("Attempting to move table from version " + fromVersion + " to " + toVersion);
+      Map<ConfigProperty, String> tableProps = new HashMap<>();
+      if (fromVersion.versionCode() < toVersion.versionCode()) {
+        // upgrade
+        while (fromVersion.versionCode() < toVersion.versionCode()) {
+          HoodieTableVersion nextVersion = HoodieTableVersion.versionFromCode(fromVersion.versionCode() + 1);
+          tableProps.putAll(upgrade(fromVersion, nextVersion, instantTime));
+          fromVersion = nextVersion;
+        }
+      } else {
+        // downgrade
+        while (fromVersion.versionCode() > toVersion.versionCode()) {
+          HoodieTableVersion prevVersion = HoodieTableVersion.versionFromCode(fromVersion.versionCode() - 1);
+          tableProps.putAll(downgrade(fromVersion, prevVersion, instantTime));
+          fromVersion = prevVersion;
+        }
+      }
+
+      // Write out the current version in hoodie.properties.updated file
+      for (Map.Entry<ConfigProperty, String> entry : tableProps.entrySet()) {
+        metaClient.getTableConfig().setValue(entry.getKey(), entry.getValue());
+      }
+      metaClient.getTableConfig().setTableVersion(toVersion);
+      createUpdatedFile(metaClient.getTableConfig().getProps());
+
+      // because for different fs the fs.rename have different action,such as:
+      // a) for hdfs : if propsFilePath already exist,fs.rename will not replace propsFilePath, but just return false
+      // b) for localfs: if propsFilePath already exist,fs.rename will replace propsFilePath, and return ture
+      // c) for aliyun ossfs: if propsFilePath already exist,will throw FileAlreadyExistsException
+      // so we should delete the old propsFilePath. also upgrade and downgrade is Idempotent
+      if (fs.exists(propsFilePath)) {
+        fs.delete(propsFilePath, false);
+      }
+      // Rename the .updated file to hoodie.properties. This is atomic in hdfs, but not in cloud stores.
+      // But as long as this does not leave a partial hoodie.properties file, we are okay.
+      fs.rename(updatedPropsFilePath, propsFilePath);
+    } catch (IOException e) {
+      throw new HoodieUpgradeDowngradeException("Error during upgrade/downgrade to version:" + toVersion, e);
+    }
+  }
+
+  private void createUpdatedFile(Properties props) throws IOException {
+    try (FSDataOutputStream outputStream = fs.create(updatedPropsFilePath)) {
+      props.store(outputStream, "Properties saved on " + new Date(System.currentTimeMillis()));
+    }
+  }
+
+  protected Map<ConfigProperty, String> upgrade(HoodieTableVersion fromVersion, HoodieTableVersion toVersion, String instantTime) {
+    if (fromVersion == HoodieTableVersion.ZERO && toVersion == HoodieTableVersion.ONE) {
+      return new ZeroToOneUpgradeHandler().upgrade(metaClient, config, context, instantTime);
```

Comment:
similar comment as above. can we avoid sending the wirteConfig as an extra argument. since we are refactoring this, wanted to take this opportunity to fix these.